sdk-rs/src/websocket_program_account_subscriber.rs

```diff
@@ -0,0 +1,184 @@
+use solana_client::rpc_config::{RpcProgramAccountsConfig, RpcAccountInfoConfig};
+use crate::types::{DataAndSlot, SdkResult};
+use log::{error, warn};
+use tokio::task::LocalSet;
+
+pub struct WebsocketProgramAccountOptions {
+    filters: Vec<RpcFilterType>,
+    commitment: CommitmentConfig,
+    encoding: UiAccountEncoding,
+}
+
+pub struct WebsocketProgramAccountSubscriber<T>
+where 
+    T: AccountDeserialize + core::fmt::Debug,
+{
+    _subscription_name: String,
+    url: String,
+    options: WebsocketProgramAccountOptions,
+    on_update: Option<Arc<dyn Fn(String, DataAndSlot<T>) -> Pin<Box<dyn Future<Output = ()>>>>>,
+    _resub_timeout_ms: Option<u64>,
+    subscribed: bool,
+}
+
+impl<T> WebsocketProgramAccountSubscriber<T>
+where 
+    T: AccountDeserialize + core::fmt::Debug,
+{
+    pub fn new(
+        subscription_name: String,
+        url: String,
+        options: WebsocketProgramAccountOptions,
+        on_update: Option<Arc<dyn Fn(String, DataAndSlot<T>) -> Pin<Box<dyn Future<Output = ()>>>>>,
+        resub_timeout_ms: Option<u64>
+    ) -> Self {
+
+        WebsocketProgramAccountSubscriber {
+            _subscription_name: subscription_name,
+            url,
+            options,
+            on_update,
+            _resub_timeout_ms: resub_timeout_ms,
+            subscribed: false, 
+        }
+    }
+
+    #[inline(always)]
+    fn decode(data: UiAccountData) -> SdkResult<T> {
+        let data_str = match data {
+            UiAccountData::Binary(encoded, _) => encoded,
+            _ => return Err(crate::types::SdkError::UnsupportedAccountData)
+        };
+
+        let decoded_data = general_purpose::STANDARD.decode(data_str)?;
+        let mut decoded_data_slice = decoded_data.as_slice();
+
+        let result = T::try_deserialize(&mut decoded_data_slice)?;
+
+        Ok(result)
+    }
+
+    pub async fn subscribe(&mut self) -> SdkResult<()> {
+        if self.subscribed {
+            return Ok(());
+        }
+        self.subscribed = true;
+        self.subscribe_ws().await?;
+
+        Ok(())
+    }
+
+    async fn subscribe_ws(&mut self) -> SdkResult<()> {
+        let account_config = RpcAccountInfoConfig {
+            commitment: Some(self.options.commitment),
+            encoding: Some(self.options.encoding),
+            ..RpcAccountInfoConfig::default()
+        };
+        let config = RpcProgramAccountsConfig {
+            filters: Some(self.options.filters.clone()),
+            account_config,
+            ..RpcProgramAccountsConfig::default()
+        };
+
+        let url = self.url.clone();
+        let mut latest_slot = 0;
+        let on_update = self.on_update.clone();
+        let local = LocalSet::new();
+
+        let (mut _subscription, receiver) = PubsubClient::program_subscribe(
+            &url,
+            &drift_program::ID,
+            Some(config)
+        )?; 
+
+        local.run_until(async move {
+
+            while let Ok(message) = receiver.recv() {
```

Comment:
I think using the blocking pubsub made this harder, with non-blocking this will be like:
```suggestion
            while let Some(message) = receiver.next.await() {
```
and can use ordinary `tokio::spawn` rather than TaskSet